drivers/flash/flash_simulator.c

```diff
@@ -0,0 +1,344 @@
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_NAME_EC))
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_NAME_EEC))
+STATS_NAME_END(flash_sim_stats);
+
+static u8_t mock_flash[FLASH_SIZE];
+static bool write_protection;
+static u64_t rand_seed;
+
+static const struct flash_driver_api flash_sim_api;
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+static int toss(int failure_rate_pm)
+{
+	rand_seed = 1103515245 * rand_seed + 12345;
+	return ((u32_t)rand_seed % 1000) < failure_rate_pm ? -EIO : 0;
+}
+#endif
+
+static int flash_range_is_valid(struct device *dev, off_t offset, size_t len)
+{
+	ARG_UNUSED(dev);
+	if (offset + len > FLASH_SIZE + CONFIG_FLASH_SIMULATOR_BASE_ADDR) {
+		return -EINVAL;
+	}
+
+	return 1;
+}
+
+static int flash_wp_set(struct device *dev, bool enable)
+{
+	ARG_UNUSED(dev);
+	write_protection = enable;
+
+	return 0;
+}
+
+static bool flash_wp_is_set(void)
+{
+	return write_protection;
+}
+
+static int flash_sim_read(struct device *dev, const off_t offset, void *data,
+			  const size_t len)
+{
+	ARG_UNUSED(dev);
+
+	if (!flash_range_is_valid(dev, offset, len)) {
+		return -EINVAL;
+	}
+
+	STATS_INC(flash_sim_stats, flash_read_calls);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+	/* randomly generate API failure */
+	int api_error = toss(CONFIG_FLASH_SIMULATOR_READ_API_FAILURE_RATE_PM);
+
+	if (api_error) {
+		STATS_INC(flash_sim_stats, flash_read_api_errors);
+		return api_error;
+	}
+#endif
+
+	memcpy(data, FLASH(offset), len);
+	STATS_INCN(flash_sim_stats, bytes_read, len);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+	/* randomly generate a hardware failure */
+	int hw_error = toss(CONFIG_FLASH_SIMULATOR_READ_HW_FAILURE_RATE_PM);
+
+	if (hw_error) {
+		/* toggle a random byte in the output buffer */
+		*((u8_t *)data + (sys_rand32_get() % len)) ^= 0xFF;
+		STATS_INC(flash_sim_stats, flash_read_hw_errors);
+	}
+#endif
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_TIMING
+	k_busy_wait(CONFIG_FLASH_SIMULATOR_MIN_READ_TIME_US);
+	STATS_INCN(flash_sim_stats, flash_read_time_us,
+		   CONFIG_FLASH_SIMULATOR_MIN_READ_TIME_US);
+#endif
+
+	return 0;
+}
+
+static int flash_sim_write(struct device *dev, const off_t offset,
+			   const void *data, const size_t len)
+{
+	ARG_UNUSED(dev);
+
+	if (!flash_range_is_valid(dev, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (offset % CONFIG_FLASH_SIMULATOR_PROG_UNIT) {
+		return -EINVAL;
+	}
+
+	if (flash_wp_is_set()) {
+		return -EACCES;
+	}
+
+	STATS_INC(flash_sim_stats, flash_write_calls);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+	/* randomly generate an API failure */
+	int api_error = toss(CONFIG_FLASH_SIMULATOR_WRITE_API_FAILURE_RATE_PM);
+
+	if (api_error) {
+		STATS_INC(flash_sim_stats, flash_write_api_errors);
+		return api_error;
+	}
+#endif
+
+	STATS_INCN(flash_sim_stats, bytes_written, len);
+	for (u32_t i = 0; i < len; i++) {
+		/* only pull bits to zero */
+		*(FLASH(offset + i)) &= *((u8_t *)data + i);
```

Comment:
This is not how most flash devices these days work. For the most part, flash devices have small writable units (typically 1-8 bytes), and these sections must only be written to once. On some devices, even writing 0xFF and then trying to write something else will fail. Other devices don't fail, but the data sheets suggest that it greatly increases wear on the devices.

I would suggest that we consider writes to anything that isn't all 0xFF to be a failure.

We might be interested, also, in enforcing a particular alignment of the writes.

Reply:
You have a point, still I would make this behavior configurable since some hardware does allow multiple writes to the same address/region.

Reply:
it's true that some flash devices can only be written once, usually (but not exclusively) the ones with ECC. This is the case for example of the STM32L4 or STM32H7. On the other hand some of them like the STM32F1 support being written multiple time, and it's even a trick often used to implement a chip (down)counter, changing one bit from 1 to 0 at each counter (dec)increment.

In short I agree with @lemrey that it should be a configurable behavior. Maybe it should not be enabled by default as most of new hardware released recently do not allow multiple write.

I also agree with @nvlsianpu that adding support for write block size would be nice to have. I guess @Laczen can confirm that it would have been useful to test NVS.

Reply:
@Laczen as @lemrey  said failure modes configuration allows  to simulate  failures expected by nor-flash devices. For now  we want to us this driver for testing stuff without failure emulation.

Reply:
That said I mean that we need to start using this, and then we can improve if this will be a case.